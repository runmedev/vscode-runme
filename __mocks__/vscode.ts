import path from 'node:path'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'

import { vi } from 'vitest'
import { URI } from 'vscode-uri'
import type { Command, Disposable } from 'vscode'


export const notebooks = {
  createNotebookController: vi.fn().mockReturnValue({
    dispose: vi.fn(),
    createNotebookCellExecution: vi.fn().mockReturnValue({ start: vi.fn(), end: vi.fn() })
  }),
  registerNotebookCellStatusBarItemProvider: vi.fn(),
  createRendererMessaging: vi.fn().mockReturnValue({
    postMessage: vi.fn(),
    onDidReceiveMessage: vi.fn().mockReturnValue({ dispose: vi.fn() })
  }),
}

export class Uri extends URI {
  static file = vi.fn(super.file)
  static parse = vi.fn(super.parse)

  static joinPath = vi.fn((uri: Uri, ...paths: string[]) => {
    return super.file(path.join(uri.fsPath, ...paths))
  })
}

export const workspace = {
  getConfiguration: vi.fn().mockReturnValue(new Map()),
  onDidChangeConfiguration: vi.fn(),
  openNotebookDocument: vi.fn().mockReturnValue({ uri: 'new notebook uri' }),
  openTextDocument: vi.fn().mockReturnValue({
    getText: vi.fn().mockReturnValue(readFileSync(join(__dirname, 'gitignore.mock'), 'utf8'))
  }),
  registerNotebookSerializer: vi.fn(),
  onDidOpenNotebookDocument: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  onDidSaveNotebookDocument: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  onDidChangeNotebookDocument: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  fs: {
    writeFile: vi.fn(),
    readFile: vi.fn().mockResolvedValue(Buffer.from('some wasm file')),
    stat: vi.fn().mockResolvedValue(1),
    createDirectory: vi.fn(),
    delete: vi.fn()
  },
  createFileSystemWatcher: vi.fn().mockReturnValue({
    onDidCreate: vi.fn(),
    onDidDelete: vi.fn()
  }),
  workspaceFolders: [ { uri: { fsPath: '/runme/workspace' } } ],
  findFiles: vi.fn().mockReturnValue([
    {
      path: 'runme/workspace/README.md',
    },
    {
      path: 'runme/workspace/src/README.md',
    },
    {
      path: 'runme/workspace/src/COMMANDS.md',
    },
    {
      path: 'runme/workspace/src/RUNME.md',
    }
  ]),
}

export const terminal = {
  show: vi.fn(),
  sendText: vi.fn(),
  dispose: vi.fn()
}

export const window = {
  createOutputChannel: vi.fn().mockReturnValue({
    appendLine: vi.fn()
  }),
  showWarningMessage: vi.fn(),
  showInformationMessage: vi.fn(),
  showErrorMessage: vi.fn(),
  createTerminal: vi.fn().mockReturnValue(terminal),
  showNotebookDocument: vi.fn(),
  showTextDocument: vi.fn(),
  onDidChangeActiveNotebookEditor: vi.fn().mockReturnValue({ dispose: vi.fn() }),
  registerTreeDataProvider: vi.fn(),
  registerUriHandler: vi.fn(),
  onDidCloseTerminal: vi.fn(),
  withProgress: vi.fn(),
  onDidChangeActiveColorTheme: vi.fn().mockReturnValue({ dispose: vi.fn() }),
}

export const tasks = {
  registerTaskProvider: vi.fn(),
  onDidEndTaskProcess: vi.fn(),
  onDidStartTaskProcess: vi.fn(),
  executeTask: vi.fn().mockResolvedValue(undefined)
}

export const commands = {
  registerCommand: vi.fn(),
  executeCommand: vi.fn(),
  getCommands: vi.fn().mockResolvedValue(['_notebook.selectKernel'])
}

export enum ViewColumn {
  Beside = 'Beside'
}

export const env = {
  clipboard: {
    writeText: vi.fn()
  },
  openExternal: vi.fn()
}

export const NotebookData = vi.fn()
export const NotebookCellData = vi.fn()
export enum NotebookCellKind {
  Code = 1,
  Markup = 2
}

export const TreeItem = vi.fn()
export const Event = vi.fn()

export enum TreeItemCollapsibleState {
  None = 0,
  Collapsed = 1,
  Expanded = 2
}

export enum FileType {
  /**
   * The file type is unknown.
   */
  Unknown = 0,
  /**
   * A regular file.
   */
  File = 1,
  /**
   * A directory.
   */
  Directory = 2,
  /**
   * A symbolic link to a file.
   */
  SymbolicLink = 64
}

export enum NotebookCellStatusBarAlignment {
  Left = 1,
  Right = 2
}

export class NotebookCellStatusBarItem {
  label: string

  alignment: NotebookCellStatusBarAlignment

  tooltip?: string

  priority?: number

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  constructor(label: string, alignment: NotebookCellStatusBarAlignment) {
    this.label = label
    this.alignment = alignment
  }
}



export class NotebookCellOutputItem {

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  static json(value: any, mime = 'text/plain'): NotebookCellOutputItem {
    return {
      mime: mime,
      data: new Uint8Array()
    }
  }

  mime: string
  data: Uint8Array
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  constructor(data: Uint8Array, mime: string) {}
}

export class NotebookCellOutput {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  constructor(public items: NotebookCellOutputItem[], public metadata?: { [key: string]: any }) {}
}

export const ProgressLocation = {
  Window: 1
}

type MessageCallback<T> = (message: T) => void
type Event<T> = (listener: MessageCallback<T>) => Disposable

export class EventEmitter<T> {
  listeners: MessageCallback<T>[] = []

  event: Event<T> = (listener) => {
    this.listeners.push(listener)

    return {
      dispose: () => {
        this.listeners = this.listeners.filter(x => x !== listener)
      }
    }
  }

  fire(data: T) {
    this.listeners.forEach(l => l(data))
  }

  async fireAsync(data: T) {
    await Promise.all(
      this.listeners.map(l => l(data))
    )
  }

  dispose() { }
}

export const languages = {
  registerCodeLensProvider: vi.fn(),
}

export class Position {
  constructor(
    public line: number,
    public character: number
  ) { }

  with(line?: number, character?: number): Position {
    return new Position(
      line ?? this.line,
      character ?? this.character,
    )
  }
}

export class Range {
  constructor(
    public start: Position,
    public end: Position,
  ) { }
}

export class CodeLens {
  constructor(
    public range: Range,
    public command?: Command
  ) { }
}

export enum TaskScope {
  /**
   * The task is a global task. Global tasks are currently not supported.
   */
  Global = 1,

  /**
   * The task is a workspace task
   */
  Workspace = 2
}

export const ShellExecution = vi.fn()
export const Task = vi.fn()
