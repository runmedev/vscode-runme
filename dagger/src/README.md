---
cwd: ../..
runme:
  id: 01HTNVRGFMWZERW6S2CZZ9E990
  version: v3
---

```sh {"id":"01HTNVRK3AJ2AT8M24TA996RCJ"}
dagger -m vscode-runme functions
```

```sh {"id":"01HTQBSZTS5M1HP3GGP4T99PT0","name":"build-binary","terminalRows":"28"}
dagger call --progress auto \
  -m golang \
  --src ../runme \
  build \
  file \
    --output /Users/sourishkrout/Projects/stateful/oss/vscode-runme/bin/runme \
    --path runme
```

```sh {"id":"01HTNZBARHB97RPQPCVQZ7PNRN","name":"build-extension","terminalRows":"25"}
dagger --progress=plain \
  call \
  with-remote \
    --remote "github.com/stateful/vscode-runme" \
    --ref "4e3007e2920299b9491ea8f4c7aa9f33cc649afc" \
  with-container \
    --binary /Users/sourishkrout/Projects/stateful/oss/vscode-runme/bin/runme \
    --presetup /Users/sourishkrout/Projects/stateful/oss/vscode-runme/presetup.sh \
  build-extension \
    --gh-token $(gh auth token)
```
