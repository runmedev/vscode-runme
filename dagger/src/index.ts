/* eslint-disable max-len */
/**
 * Build the Runme VS Code extension end-to-end.
 *
 */

// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { dag, Container, File, Directory, object, func, field } from '@dagger.io/dagger'

@object()
// eslint-disable-next-line @typescript-eslint/no-unused-vars
class VscodeRunme {
  /**
   * The working repository directory for the VscodeRunme instance.
   */
  @field()
  dir: Directory

  /**
   * The base container being used for building the extension
   */
  @field()
  container: Container

  /**
   * Build from remote git repository.
   * @param remote - Valid git remote url, aka repo to clone
   * @param ref - Branch, tag, or commit to checkout
   * @returns The modified VscodeRunme instance.
   */
  @func()
  withRemote(remote: string, ref: string): VscodeRunme {
    this.dir = dag
    .git(`https://${remote}.git`)
    .ref(ref)
    .tree()

    return this
  }

  /**
   * Sets up the container for the VscodeRunme instance.
   * @param binary - Optional kernel binary file to be added to the container.
   * @param presetup - Optional presetup (for dependencies) file to be added to the container.
   * @returns The modified VscodeRunme instance.
   */
  @func()
  withContainer(binary?: File, presetup?: File): VscodeRunme {
    this.container = dag
      .container()
      .from('node:18')
      .withEnvVariable('EXTENSION_NAME', 'runme')
      .withFile('/usr/local/bin/runme', binary)
      .withFile('/usr/local/bin/presetup', presetup)
      .withEntrypoint([])
      .withMountedDirectory('/mnt/vscode-runme', this.dir)
      .withWorkdir('/mnt/vscode-runme')
      .withExec('bash /usr/local/bin/presetup'.split(' '))

      return this
  }

  /**
   * Sets up the container for the VscodeRunme instance.
   * @param ghToken - Valid GitHub access token for API access.
   * @returns The modified VscodeRunme instance.
   */
  @func()
  async buildExtension(ghToken: string): Promise<string> {
    return this.container
      .withEnvVariable('GITHUB_TOKEN', ghToken)
      .withExec('runme run setup build'.split(' ')).
      stdout()
  }
}
