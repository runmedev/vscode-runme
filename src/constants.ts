/**
 * Note: this file is used within Node.js and Browser environment.
 * Only export cross compatible objects here.
 */

export enum OutputType {
  vercel = 'stateful.runme/vercel-stdout',
  deno = 'stateful.runme/deno-stdout',
  outputItems = 'stateful.runme/output-items',
  annotations = 'stateful.runme/annotations',
  error = 'stateful.runme/error'
}

export enum ClientMessages {
  infoMessage = 'common:infoMessage',
  errorMessage = 'common:errorMessage',
  update = 'deno:deploymentUpdate',
  deployed = 'deno:finishedDeployment',
  promote = 'deno:promoteDeployment',
  prod = 'vercel:promotePreview',
  mutateAnnotations = 'annotations:mutate'
}

// [pretty print, languageId, destination]
export const LANGUAGES = new Map([
  ['Assembly', 'asm', 'sh'],
  ['Batchfile', 'bat', 'sh'],
  ['C', 'c', undefined],
  ['C#', 'cs', undefined],
  ['C++', 'cpp', undefined],
  ['Clojure', 'clj', undefined],
  ['CMake', 'cmake', 'sh'],
  ['COBOL', 'cbl', 'null'],
  ['CoffeeScript', 'coffee', 'html'],
  ['CSS', 'css', undefined],
  ['CSV', 'csv', undefined],
  ['Dart', 'dart', undefined],
  ['DM', 'dm', undefined],
  ['Dockerfile', 'dockerfile', 'sh'],
  ['Elixir', 'ex', undefined],
  ['Erlang', 'erl', undefined],
  ['Fortran', 'f90', undefined],
  ['Go', 'go', undefined],
  ['Groovy', 'groovy', undefined],
  ['Haskell', 'hs', undefined],
  ['HTML', 'html', undefined],
  ['INI', 'ini', undefined],
  ['Java', 'java', undefined],
  ['JavaScript', 'js', undefined],
  ['JSON', 'json', undefined],
  ['Julia', 'jl', undefined],
  ['Kotlin', 'kt', undefined],
  ['Lisp', 'lisp', undefined],
  ['Lua', 'lua', undefined],
  ['Makefile', 'makefile', 'sh'],
  ['Markdown', 'md', undefined],
  ['Matlab', 'matlab', undefined],
  ['Objective-C', 'mm', undefined],
  ['OCaml', 'ml', undefined],
  ['Pascal', 'pas', undefined],
  ['Perl', 'pm', undefined],
  ['PHP', 'php', undefined],
  ['PowerShell', 'ps1', undefined],
  ['Prolog', 'prolog', undefined],
  ['Python', 'py', undefined],
  ['R', 'r', undefined],
  ['Ruby', 'rb', undefined],
  ['Rust', 'rs', undefined],
  ['Scala', 'scala', undefined],
  ['Shell', 'sh', 'sh'],
  ['SQL', 'sql', undefined],
  ['Swift', 'swift', undefined],
  ['TeX', 'tex', undefined],
  ['TOML', 'toml', undefined],
  ['TypeScript', 'ts', undefined],
  ['Verilog', 'v', undefined],
  ['Visual Basic', 'vba', undefined],
  ['XML', 'xml', undefined],
  ['YAML', 'yaml', undefined],
].map(([, source, dest]) => [source, dest]))
