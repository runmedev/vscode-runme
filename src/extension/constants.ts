import os from 'node:os'

export const PLATFORM_OS = os.platform()
export const DENO_ACCESS_TOKEN_KEY = 'DENO_ACCESS_TOKEN'
export const DENO_PROJECT_NAME_KEY = 'DENO_PROJECT_NAME'

export const DEFAULT_ENV = {
  RUNME_TASK: 'true',
  PATH: process.env.PATH || ''
}
export const ENV_STORE = new Map<string, string>(
  Object.entries(DEFAULT_ENV)
)

export const BOOTFILE = '.runme_bootstrap'
