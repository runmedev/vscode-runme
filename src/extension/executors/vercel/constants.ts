export const VERCEL_DIR = '.vercel'
