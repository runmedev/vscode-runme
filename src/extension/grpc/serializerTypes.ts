export * from '@buf/stateful_runme.community_timostamm-protobuf-ts/runme/parser/v1/parser_pb'
