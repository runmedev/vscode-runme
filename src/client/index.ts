import type { ActivationFunction, RendererContext } from 'vscode-notebook-renderer'

import { OutputType, RENDERERS } from '../constants'
import type { CellOutput } from '../types'

import { setContext } from './utils'
import './components'

// ----------------------------------------------------------------------------
// This is the entrypoint to the notebook renderer's webview client-side code.
// This contains some boilerplate that calls the `render()` function when new
// output is available. You probably don't need to change this code; put your
// rendering logic inside of the `render()` function.
// ----------------------------------------------------------------------------

export const activate: ActivationFunction = (context: RendererContext<void>) => {
  setContext(context)
  return {
    renderOutputItem(outputItem, element) {
      const payload: CellOutput = outputItem.json()

      switch (payload.type) {
        case OutputType.vercel:
          if (payload.output.error) {
            renderError(payload.output.error)
            break
          }

          const vercelElem = document.createElement(RENDERERS.VercelOutput)
          vercelElem.setAttribute('content', JSON.stringify(payload.output))
          element.appendChild(vercelElem)
          break
        case OutputType.deno:
          if (payload.output?.error) {
            renderError(payload.output.error)
            break
          }

          const deno = payload.output || {}
          const denoElem = document.createElement(RENDERERS.DenoOutput)

          denoElem.setAttribute('state', JSON.stringify(deno))

          element.appendChild(denoElem)
          break
        case OutputType.outputItems:
          const content = decodeURIComponent(escape(window.atob(payload.output.content)))
          /**
           * shell output
           */
          const shellElem = document.createElement(RENDERERS.ShellOutput)
          shellElem.innerHTML = content
          element.appendChild(shellElem)
          /**
           * output items, e.g. copy to clipboard
           */
          const outputItemElem = document.createElement(RENDERERS.ShellOutputItems)
          outputItemElem.setAttribute('uuid', payload.output.uuid)
          outputItemElem.setAttribute('content', content)
          element.appendChild(outputItemElem)
          break
        case OutputType.annotations:
          const annoElem = document.createElement(RENDERERS.EditAnnotations)
          annoElem.setAttribute('annotations', JSON.stringify(payload.output.annotations ?? []))
          annoElem.setAttribute('validationErrors', JSON.stringify(payload.output.validationErrors ?? []))
          element.appendChild(annoElem)
          break
        case OutputType.terminal:
          const terminalElement = document.createElement(RENDERERS.TerminalView)
          terminalElement.setAttribute('uuid', payload.output['runme.dev/uuid'])
          terminalElement.setAttribute('terminalFontFamily', payload.output.terminalFontFamily)
          terminalElement.setAttribute('terminalFontSize', payload.output.terminalFontSize.toString())

          if (payload.output.initialRows !== undefined) {
            terminalElement.setAttribute('initialRows', payload.output.initialRows.toString())
          }

          if (payload.output.content !== undefined) {
            terminalElement.setAttribute('initialContent', payload.output.content)
          }

          element.appendChild(terminalElement)
          break
        case OutputType.error:
          renderError(payload.output)
          break
        case OutputType.github:
          const githubElement = document.createElement(RENDERERS.GitHubWorkflowViewer)
          if (payload.output?.content) {
            githubElement.setAttribute('state', JSON.stringify(payload.output))
          }
          element.appendChild(githubElement)
          break
        default: element.innerHTML = 'No renderer found!'
      }

      function renderError(message: string) {
        element.innerHTML = `⚠️ ${message}`
      }
    },
    disposeOutputItem(/* outputId */) {
      // Do any teardown here. outputId is the cell output being deleted, or
      // undefined if we're clearing all outputs.
    }
  }
}
