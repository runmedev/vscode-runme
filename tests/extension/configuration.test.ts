import path from 'node:path'

import { suite, test, expect, vi, beforeEach, afterEach } from 'vitest'
import { Uri, workspace } from 'vscode'

import {
  getRunmePanelIdentifier,
  getRunmeAppUrl,
  getPortNumber,
  enableServerLogs,
  getBinaryPath,
  getServerConfigurationValue,
  getTLSDir,
  getNotebookTerminalConfigurations,
  getCodeLensEnabled,
  getCLIUseIntegratedRunme,
} from '../../src/utils/configuration'
import { SERVER_PORT } from '../../src/constants'
import { RunmeIdentity } from '../../src/extension/grpc/serializerTypes'

vi.mock('../../src/extension/grpc/client', () => ({}))
vi.mock('../../src/extension/grpc/runnerTypes', () => ({}))

const FAKE_UNIX_EXT_PATH = '/Users/user/.vscode/extension/stateful.runme'
const FAKE_WIN_EXT_PATH = 'C:\\Users\\.vscode\\extensions\\stateful.runme'

vi.mock('path', async (origModFactory) => {
  const origMod = await origModFactory<typeof path>()
  const p = {
    ...origMod,
    join: vi.fn(origMod.join),
    isAbsolute: vi.fn(origMod.isAbsolute),
  }
  return { ...p, default: p }
})

vi.mock('vscode', async () => {
  const mocked = await import('../../__mocks__/vscode')
  const SETTINGS_MOCK: {
    port: number | string | undefined
    binaryPath: string | undefined
    enableLogger: string | boolean | undefined
    tlsDir: string | undefined
    baseDomain: string | undefined
    'runme.cloud': string | undefined
  } = {
    port: undefined,
    binaryPath: undefined,
    enableLogger: undefined,
    tlsDir: undefined,
    baseDomain: undefined,
    'runme.cloud': undefined,
  }

  return {
    ...mocked,
    workspace: {
      getConfiguration: vi.fn().mockReturnValue({
        update: (configurationName: string, val: unknown) => {
          SETTINGS_MOCK[configurationName] = val
        },
        get: (configurationName) => {
          return SETTINGS_MOCK[configurationName]
        },
      }),
    },
    Uri: mocked.Uri,
  }
})

vi.mock('vscode-telemetry')

suite('Configuration', () => {
  test('should get nullish from font family', () => {
    expect(getNotebookTerminalConfigurations()).toMatchSnapshot()
    workspace.getConfiguration().update('fontFamily', 'Fira Code')
    expect(getNotebookTerminalConfigurations().fontFamily).toStrictEqual('Fira Code')

    // example fails because "line" is not valid
    workspace.getConfiguration().update('cursorStyle', 'line')
    expect(getNotebookTerminalConfigurations().cursorStyle).toStrictEqual(undefined)

    workspace.getConfiguration().update('cursorStyle', 'underline')
    expect(getNotebookTerminalConfigurations().cursorStyle).toStrictEqual('underline')
  })

  test('should default to a valid port number', () => {
    const portNumber = getPortNumber()
    expect(portNumber).toStrictEqual(7863)
  })

  test('should use a valid specified port number', () => {
    const portNumber = getPortNumber()
    expect(portNumber).toStrictEqual(SERVER_PORT)
  })

  test('should disable server logs with an invalid value', () => {
    workspace.getConfiguration().update('enableLogger', undefined)
    const path = enableServerLogs()
    expect(path).toBeFalsy()
  })

  test('should disable server logs with an invalid string', () => {
    workspace.getConfiguration().update('enableLogger', 'true')
    const path = enableServerLogs()
    expect(path).toBeFalsy()
  })

  test('should get default TLS dir by default', () => {
    workspace.getConfiguration().update('tlsDir', undefined)
    expect(getTLSDir(Uri.file('/ext/base'))).toBe(Uri.file('/ext/base/tls').fsPath)
  })

  test('should get set TLS dir if set', () => {
    workspace.getConfiguration().update('tlsDir', '/tmp/runme/tls')
    expect(getTLSDir(Uri.file('/ext/base'))).toBe('/tmp/runme/tls')
  })

  test('getServerConfigurationValue should default to undefined binaryPath', () => {
    workspace.getConfiguration().update('binaryPath', undefined)

    expect(getServerConfigurationValue<string | undefined>('binaryPath', undefined)).toStrictEqual(
      undefined,
    )
  })

  test('getServerConfigurationValue should give proper binaryPath if defined', () => {
    workspace.getConfiguration().update('binaryPath', '/binary/path')

    expect(getServerConfigurationValue<string | undefined>('binaryPath', undefined)).toStrictEqual(
      '/binary/path',
    )
  })

  test('getServerConfigurationValue should give default persist identity', () => {
    expect(
      getServerConfigurationValue<number>('persistIdentity', RunmeIdentity.UNSPECIFIED),
    ).toStrictEqual(RunmeIdentity.ALL)
  })

  test('getCodeLensEnabled should return true by default', () => {
    expect(getCodeLensEnabled()).toStrictEqual(true)
  })

  test('getCLIUseIntegratedRunme should return false by default', () => {
    expect(getCLIUseIntegratedRunme()).toStrictEqual(false)
  })

  suite('posix', () => {
    beforeEach(() => {
      workspace.getConfiguration().update('binaryPath', undefined)
    })

    afterEach(() => {
      workspace.getConfiguration().update('binaryPath', undefined)
    })

    test('should default to a valid binaryPath', () => {
      const binary = getBinaryPath(Uri.file(FAKE_UNIX_EXT_PATH), 'linux')
      expect(binary.fsPath).toStrictEqual('/Users/user/.vscode/extension/stateful.runme/bin/runme')
    })

    test('should default to a valid relative binaryPath when specified', () => {
      workspace.getConfiguration().update('binaryPath', 'newBin')
      // @ts-expect-error readonly
      workspace.workspaceFolders = [{ uri: Uri.file('/Users/user/Projects/project') }]
      const binary = getBinaryPath(Uri.file(FAKE_UNIX_EXT_PATH), 'linux')
      expect(binary.fsPath).toStrictEqual('/Users/user/Projects/project/newBin')
    })

    test('should default to a valid absolute binaryPath when specified', () => {
      workspace.getConfiguration().update('binaryPath', '/opt/homebrew/bin/runme')
      const binary = getBinaryPath(Uri.file(FAKE_UNIX_EXT_PATH), 'linux')
      expect(binary.fsPath).toStrictEqual('/opt/homebrew/bin/runme')
    })

    test('should use runme for non-windows platforms', () => {
      workspace.getConfiguration().update('binaryPath', '/opt/homebrew/bin/runme')
      const binary = getBinaryPath(Uri.file(FAKE_UNIX_EXT_PATH), 'darwin')
      expect(binary.fsPath).toStrictEqual('/opt/homebrew/bin/runme')
    })
  })

  suite('win32', () => {
    beforeEach(() => {
      path.join = vi.fn(path.win32.join)
      path.isAbsolute = vi.fn(path.win32.isAbsolute)
    })

    afterEach(() => {
      vi.mocked(path.join).mockRestore()
      vi.mocked(path.isAbsolute).mockRestore()
    })

    test('should default to a valid binaryPath exe on windows', () => {
      const binary = getBinaryPath(Uri.file(FAKE_WIN_EXT_PATH), 'win32')
      expect(binary.fsPath).toStrictEqual(
        'c:\\Users\\.vscode\\extensions\\stateful.runme\\bin\\runme.exe',
      )
    })

    test('should use runme.exe for windows platforms with absolute path', () => {
      workspace.getConfiguration().update('binaryPath', 'C:\\custom\\path\\to\\bin\\runme.exe')

      const binary = getBinaryPath(Uri.file(FAKE_WIN_EXT_PATH), 'win32')
      expect(binary.fsPath).toStrictEqual('c:\\custom\\path\\to\\bin\\runme.exe')
    })

    test('should use runme.exe for windows platforms with relative path', () => {
      workspace.getConfiguration().update('binaryPath', 'newBin.exe')
      // @ts-expect-error readonly
      workspace.workspaceFolders = [{ uri: Uri.file('c:\\Users\\Projects\\project') }]
      const binary = getBinaryPath(Uri.file(FAKE_WIN_EXT_PATH), 'win32')
      expect(binary.fsPath).toStrictEqual('c:\\Users\\Projects\\project\\newBin.exe')
    })
  })

  suite('app domain resolution', () => {
    test('should return URL for api with subdomain', () => {
      const url = getRunmeAppUrl(['api'])
      expect(url).toStrictEqual('https://api.runme.dev/')
    })

    test('should return URL for api with deep subdomain', () => {
      const url = getRunmeAppUrl(['l4', 'l3', 'api'])
      expect(url).toStrictEqual('https://l4.l3.api.runme.dev/')
    })

    test('should return URL without subdomain', () => {
      const url = getRunmeAppUrl([])
      expect(url).toStrictEqual('https://runme.dev/')
    })

    test('should return URL without subdomain', () => {
      const url = getRunmeAppUrl([])
      expect(url).toStrictEqual('https://runme.dev/')
    })

    test('should allow api URL with http for 127.0.0.1', async () => {
      workspace.getConfiguration().update('baseDomain', '127.0.0.1')
      const url = getRunmeAppUrl(['api'])
      expect(url).toStrictEqual('http://127.0.0.1:4000/')
    })

    test('should allow app URL with http for localhost', async () => {
      workspace.getConfiguration().update('baseDomain', 'localhost')
      const url = getRunmeAppUrl(['app'])
      expect(url).toStrictEqual('http://localhost:4001/')
    })

    test('should allow app URL with http for localhost without subdomain', async () => {
      workspace.getConfiguration().update('baseDomain', 'localhost')
      const url = getRunmeAppUrl([])
      expect(url).toStrictEqual('http://localhost/')
    })

    test('should allow specific app URL for remote dev returning staging-based domains', async () => {
      workspace.getConfiguration().update('baseDomain', 'http://localhost:4001')
      const app = getRunmeAppUrl(['app'])
      expect(app).toStrictEqual('http://localhost:4001')
      const api = getRunmeAppUrl(['api'])
      expect(api).toStrictEqual('https://api.staging.runme.dev/')
    })
  })

  suite('app panel custom assigment', () => {
    test('should return key value for default', () => {
      const id = getRunmePanelIdentifier('runme.cloud')
      expect(id).toStrictEqual('runme.cloud')
    })

    test('should return respective value for key', () => {
      workspace.getConfiguration().update('runme.cloud', 'runme.another')
      const id = getRunmePanelIdentifier('runme.cloud')
      expect(id).toStrictEqual('runme.another')
    })
  })
})
