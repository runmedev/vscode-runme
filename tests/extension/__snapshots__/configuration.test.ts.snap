// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Configuration > should get nullish from font family 1`] = `
{
  "backgroundTask": true,
  "cursorBlink": undefined,
  "cursorStyle": "bar",
  "cursorWidth": undefined,
  "fontFamily": undefined,
  "fontSize": undefined,
  "interactive": true,
  "nonInteractive": false,
  "rows": 10,
  "scrollback": undefined,
  "smoothScrollDuration": undefined,
}
`;
